File: leetCode/双指针/15三数之和.go
package main

import (
	"fmt"
)

func sort(nums []int) []int {
	if len(nums) == 0 {
		return nil
		// return []int{}
	}
	val := nums[0]
	leftArray := []int{}
	rightArray := []int{}
	for i := 1; i < len(nums); i++ {
		if nums[i] < val {
			leftArray = append(leftArray, nums[i])
		} else {
			rightArray = append(rightArray, nums[i])
		}
	}
	return append(append(sort(leftArray), val), sort(rightArray)...)
}
func threeSum(nums []int) [][]int {
	result := [][]int{}
	sortNums := sort(nums)
	for i := 0; i < len(sortNums); i++ {
		leftP := i + 1
		rightP := len(sortNums) - 1
		if i > 0 && sortNums[i-1] == sortNums[i] {
			continue
		}
		for leftP < rightP {
			sum := sortNums[i] + sortNums[leftP] + sortNums[rightP]
			if sum < 0 {
				leftP++
			} else if sum > 0 {
				rightP--
			} else {
				item := []int{sortNums[i], sortNums[leftP], sortNums[rightP]}
				result = append(result, item)
				leftP++
				rightP--

				for leftP < rightP && sortNums[leftP] == sortNums[leftP-1] {
					leftP++
				}

				for leftP < rightP && sortNums[rightP] == sortNums[rightP+1] {
					rightP--
				}
			}
		}
	}
	return result
}
func main() {
	fmt.Println(threeSum([]int{-1, 0, 1, 2, -1, -4}))
}

File: leetCode/2.双指针/15三数之和.js
/**
 * @param {number[]} nums
 * @return {number[][]} -4 -1 -1 0 1 2
 */
var threeSum = function(nums) { 
    const answer = []
    nums = sort(nums)
    for(let i = 0; i < nums.length; i++) {
        let pl = i + 1
        let pr = nums.length - 1
        if (i > 0 && nums[i - 1] === nums[i]) continue
        while(pl < pr) {
            const sum = nums[i] + nums[pl] + nums[pr]
            if (sum < 0) {
                while(pl < pr && nums[pl] === nums[++pl]);
            }
            if (sum > 0) {
                while(pl < pr && nums[pr] === nums[--pr]);
            }
            if (sum === 0) {
                answer.push([nums[i], nums[pl], nums[pr]])
                while(pl < pr && nums[pl] === nums[++pl]);
                while(pl < pr && nums[pr] === nums[--pr]);
            }
        }
    }
    return answer
};
var sort = (array) => {
    if (array.length === 0) {
        return []
    }
    let targetI = 0
    let leftArray = []
    let rightArray = []
    for (let i = 1; i < array.length; i++) {
        if (array[i] <= array[targetI]) {
            leftArray.push(array[i])
        } else {
            rightArray.push(array[i])
        }
    }
    return [...sort(leftArray), array[targetI], ...sort(rightArray)]
}
console.log(threeSum([-1,0,1,2,-1,-4]))